index.bs

```diff
@@ -1469,3 +1479,446 @@ The [=remote end steps=] with |command parameters| are:
+         followed by a colon ("<code>:</code>") followed by a type name.
+   </dl>
+
+ 1. Let |origin| be the [=serialization of an origin=] given |environment settings|'s |origin|.
+
+ 1. Let |realm info| be a map matching the <code>RealmInfo</code> production,
+    with the <code>realm</code> field set to |realm id|, the <code>type</code>
+    field set to |type| and the <code>origin</code> field set to |origin|.
+
+ 1. Return |realm info|
+
+</div>
+
+### Commands ### {#module-script-commands}
+
+#### getRealms ####  {#command-script-getrealms}
+
+The <dfn export for=commands>getRealms</dfn> command returns a list of all
+realms, optionally filtered to realms of a specific type, or to the realm
+associated with the document currently loaded in a specified browsing context.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      GetRealmsCommand = {
+        method: "script.getRealms",
+        params: GetRealmsParameters
+      }
+
+      GetRealmsParameters = {
+        ?context: BrowsingContext,
+        ?type: RealmType,
+      }
+      </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl local-cddl">
+        GetRealmsResult = {
+          realms: [*RealmInfo]
+        }
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for script.getRealms">
+The [=remote end steps=] with |command parameters| are:
+
+  1. Let |environment settings| be a list of all the [=environment settings objects=]
+     that have their [=execution ready flag=] set.
+
+  1. If |command parameters| contains <code>context</code>:
+
+    1. Let |context| be the result of [=trying=] to [=get a browsing context=]
+       with |command parameters|["<code>context</code>"].
+
+    1. Let |document| be |context|'s [=active document=].
+
+    1. Let |context environment settings| be a list.
+
+    1. For each |settings| of |environment settings|:
+
+      1. If any of the following conditions hold:
+
+        * The [=responsible document=] of |settings| is |document|
+
+        * The [=Realm/global object=] specified by |settings| is a
+          {{WorkerGlobalScope}} with |document| in its [=owner set=]
+
+        Append |settings| to |context environment settings|.
+
+    1. Set |environment settings| to |context environment settings|.
+
+  1. Let |realms| be a list.
+
+  1. For each |settings| of |environment settings|:
+
+    1. Let |realm info| be the result of [=get the realm info=] given |settings|
+
+    1. If |realm info| is not null, append |realm info| to |realms|.
+
+  1. Let |body| be a map matching the <code>GetRealmsResult</code> production,
+     with the <code>realms</code> field set to |realms|.
+
+  1. Return [=success=] with data |body|.
+
+
+Issue: Extend this to also allow realm parents e.g. for nested workers? Or get all ancestor workers.
+
+</div>
+
+### Events ### {#module-script-events}
+
+#### realmCreated #### {#event-realm-created}
```

Comment:
What about a `realmDestroyed` event? We could potentially fire this during the steps to [discard a document](https://html.spec.whatwg.org/#garbage-collection-and-browsing-contexts) for "window" and "worklet" realms, and during [terminate a worker](https://html.spec.whatwg.org/#terminate-a-worker) for the worker types.

Reply:
I added this. Working out where it's should be invoked is hard because that's intentionally unexposed to content...